### PR TITLE
Issue 1992

### PR DESCRIPTION
**Pull request by JustinKyleJames**

### CODE REVIEW

File: packaging/resource_suite_s3_cache.py [hunk 1]
```diff
@@ -90,10 +90,13 @@ def setUp(self):
                               region=self.s3region,
                               secure=False)
 
-        self.s3bucketname = 'irods-ci-' + distro_str + datetime.datetime.utcnow().strftime('-%Y-%m-%d.%H-%M-%S-%f-')
-        self.s3bucketname += ''.join(random.choice(string.letters) for i in xrange(10))
-        self.s3bucketname = self.s3bucketname[:63].lower() # bucket names can be no more than 63 characters long
-        s3_client.make_bucket(self.s3bucketname, location=self.s3region)
+        try:
+            self.s3bucketname = self.static_bucket_name
+        except AttributeError:
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_cache.py:95]: done

File: s3/s3_operations.cpp [hunk 2]
```diff
@@ -761,39 +763,53 @@ namespace irods_s3 {
...
                                 "Try again later.");
                         addRErrorMsg( &_ctx.comm()->rError, 0, result.result().c_str());
-                        return(PASS(result));
-                    }
+                        break;
+
+                    default:
+
+                        // invalid object status - should not happen
+                        result = ERROR(REPLICA_IS_BEING_STAGED, "Invalid S3 object status detected.");
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/s3_operations.cpp:804]: done

File: s3/s3_transport/src/s3_transport.cpp [hunk 3]
```diff
@@ -39,8 +39,100 @@
...
+        rodsLog(LOG_DEBUG, "%s:%d (%s) [[%lu]] [key=%s][manager.status=%s]\n", __FILE__, __LINE__,
+                __FUNCTION__, thread_id, object_key.c_str(), S3_get_status_name(upload_manager.status));
+
+        if (upload_manager.status != S3StatusOK) {
+
+            rodsLog(LOG_ERROR, "%s:%d (%s) [[%lu]] S3_restore_object returned error [status=%s][object_key=%s].\n",
+                    __FILE__, __LINE__, __FUNCTION__, thread_id,
+                    S3_get_status_name(upload_manager.status), object_key.c_str());
+
+            return ERROR(REPLICA_IS_BEING_STAGED, "Object is in glacier and scheduling restoration failed.");
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/s3_transport/src/s3_transport.cpp]: done

File: s3/libirods_s3.cpp [hunk 4]
```diff
@@ -209,7 +214,7 @@ void s3_sleep(
     randMutex.lock();
     int random = rand();
     randMutex.unlock();
-    int sleep_time = (int)((((double)random / (double)RAND_MAX) + 1) * .5 * _s); // sleep between _s/2 and _s 
+    int sleep_time = (int)((((double)random / (double)RAND_MAX) + 1) * .5 * _s); // sleep between _s/2 and _s
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/libirods_s3.cpp:221]: done

File: CMakeLists.txt [hunk 5]
```diff
@@ -24,7 +24,8 @@ macro(IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIE
   IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH(${DEPENDENCY_NAME} ${DEPENDENCY_SUBDIRECTORY})
   list(APPEND IRODS_PACKAGE_DEPENDENCIES_LIST irods-externals-${DEPENDENCY_SUBDIRECTORY})
 endmacro()
-IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(S3 libs34e684077-0)
+#IRODS_MACRO_CHECK_DEPENDENCY_SET_FULLPATH_ADD_TO_IRODS_PACKAGE_DEPENDENCIES_LIST(S3 libs3debug-0)  # with verbosity
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply CMakeLists.txt]: Added the debug version if the -DCMAKE_BUILD_TYPE is set to 'Debug'.

File: s3/s3_transport/include/s3_transport.hpp [hunk 6]
```diff
@@ -1209,14 +1219,35 @@ namespace irods::experimental::io::s3_transport
...
+                    } else {
+                        this->set_error(ERROR(REPLICA_IS_BEING_STAGED, "Object is in glacier and is currently being restored.  "
+                                    "Try again later."));
+                    }
+
+                    return_value = false;
+                    return;
+                }
+
+		        if (object_status == object_s3_status::IN_S3 && this->download_to_cache_) {
```
[the remark this answers is not in the document]
Comment (JustinKyleJames) [reply s3/s3_transport/include/s3_transport.hpp:1251]: done

**Comment by trel** [comment]
i think we're ready to squash to three commits?... one for 1992, one for 1996, and one for libs3#12?

**Comment by trel** [comment]
i think we're good to go - please # em.

**Pull request merged**